### ISSUE TITLE
[Feature Request]: curve: layout of animate_curve and animate_3D_curve

### ISSUE THREAD

**Issue by martin-henz** [source-academy/modules#218]
# Feature

Currently we have:

<img width="1165" alt="Screenshot 2023-09-03 at 10 18 35 PM" src="https://github.com/source-academy/modules/assets/30220768/0eb49de8-e408-4052-b496-64366259b6e3">

Note the wasted space:

<img width="1165" alt="Screenshot 2023-09-03 at 10 18 35 PM copy" src="https://github.com/source-academy/modules/assets/30220768/7d508442-7bd3-4105-a789-cf923da15246">

The two items, controls and display, should be arranged vertically instead of horizontally. That would lead to a much more efficient use of screen space.

**Comment by martin-henz** [comment]
Much better now! Thanks @sayomaki 

A small request: At the moment, the width of the animation is limited to the width of the controls. Could  we change that so that the animation fills as much of the tab space as possible?

At the moment, we have:

<img width="820" alt="Screenshot 2023-09-04 at 2 12 52 PM" src="https://github.com/source-academy/modules/assets/30220768/6e07081f-96aa-498d-bbcb-a918825b76fc">

**Comment by sayomaki** [comment]
I did consider that (stretching the sliders) when designing the change, but it would result in a pretty long bar, especially for those with really large screens. 

Therefore, I propose that we can let the slider expand in tandem the page width, but only up to a certain length relative to viewport width, so as to not completely stretch across the whole available length (considering how the rendered model only occupies a fixed maximum length proportional to viewport width, as seen in your screenshot above). 

Alternatively, we can adjust the size of the model render output as well (together with the slider length), but that will result in the canvas size being inconsistent with all the other modules.

**Comment by martin-henz** [comment]
I think we are coming to a consensus:

The model widget should expand proportionally to maximize use of available space below the sliders.

Regarding the question whether the sliders should always expand to occupy the whole width or whether there should be a maximal width: I don't have a particular preference. If you find a maximal width more pleasing, please go ahead.

**Comment by Cloud7050** [comment]
> I think we are coming to a consensus:
> 
> The model widget should expand proportionally to maximize use of available space below the sliders.
> 
> Regarding the question whether the sliders should always expand to occupy the whole width or whether there should be a maximal width: I don't have a particular preference. If you find a maximal width more pleasing, please go ahead.

I have been taking a deeper look at this issue and I think the second part of the enhancement has more in the way of complexity than would initially appear. I am still compiling my findings and will post them shortly.

**Comment by Cloud7050** [comment]
> I did consider that (stretching the sliders) when designing the change, but it would result in a pretty long bar, especially for those with really large screens.
> 
> Therefore, I propose that we can let the slider expand in tandem the page width, but only up to a certain length relative to viewport width, so as to not completely stretch across the whole available length (considering how the rendered model/animation only occupies a fixed maximum length proportional to viewport width, as seen in your screenshot above).
> 
> Alternatively, we can adjust the size of the model render output/animation as well (together with the slider length), but that will result in the animation canvas size being inconsistent with all the other modules.

If I am understanding Prof Martin correctly, he feels that the width of the controls are fine, but would like the canvas to be even larger so as to use up the remaining tab space available. This is as opposed to the notion of keeping the existing canvas size shown in the screenshot, while making only the sliders wider to use up more tab space.

Currently, the maximum canvas width is not limited by the width of the controls, rather, the controls flex dynamically to match the width of the canvas. (Although, the minimum width of the canvas is limited by how skinny flexbox allows the controls to be.) There appear to be challenges involved with making the canvas larger than in the screenshot which cause this enhancement to no longer be a quick/simple change.

Curves currently seem to render at a fixed 512x512 resolution regardless of actual canvas size. Curves without controls (such as non-animated, non-3D curves) flex freely as there are no sibling controls in the same parent `div`, but the rendering is only done once, so a larger canvas element results in a blurry image:

![image](https://github.com/source-academy/modules/assets/44526554/96cc1734-80c3-41ea-9901-0281b785adfb)

As for curves with controls (where interacting with the controls results in a redraw), when the canvas is smaller than 512x512, it renders at a higher resolution than necessary, which is a bit wasteful. Meanwhile, the canvas is technically allowed to stretch larger than that size by way of its CSS styling, but the corresponding fixed width and height properties together with how flexbox works would appear to prevent it from doing so. **I believe this is the crux of the problem.** Even if we were to *force* the canvas to appear at a larger size by somehow modifying the canvas properties on the fly, the renderer is still operating at the fixed lower resolution, and this either causes a blurry image or *undefined behaviour*:

![image](https://github.com/source-academy/modules/assets/44526554/f55f9b24-ad24-4710-9f92-6749bd441d30)

![image](https://github.com/source-academy/modules/assets/44526554/785bb9d3-1083-4edf-96fc-afd0a9b01801)

The current CSS styling of the canvases come from the creation of CSG, where I experimented with various canvas sizes and how they could adapt to different tab sizes on different platforms. I believe curve/rune canvas styling used to be different, but with Lee Yi's additions to those modules such as module state, animated/3D and multiple tabbed renders, the same CSS styling as CSG was adopted. In this way, I think the concern of inconsistency with other modules is addressed.

For CSG, the canvas will expand to take up as much of the tab space as possible while avoiding becoming ridiculously large.
- For mobile views, the screen is skinny yet tall. CSS `maxWidth` will be `70vh`
- For web views, the window is typically wide but short. The CSS `maxWidth` usually in use will be `30vw`

Taking those together, the relevant CSS styling for the `div` containing the canvas is `maxWidth: 'max(70vh, 30vw)'` (as seen in `canvas_holder.tsx`). Below are some images illustrating the dynamic sizing and capped width:

Conventional usage (use up all available tab space):

![image](https://github.com/source-academy/modules/assets/44526554/95f2d794-a176-4c74-ae81-747995be3ea1)

![image](https://github.com/source-academy/modules/assets/44526554/5613175c-7d6f-41e4-94a9-f4cc0c07d1ff)

Edge cases (cap the width to avoid the canvas being too large):

![image](https://github.com/source-academy/modules/assets/44526554/ef971ef9-e5a2-4c62-95bd-4d7e01222f2f)

![image](https://github.com/source-academy/modules/assets/44526554/74590fd6-cde2-4afd-bcf4-f4852c04a531)

CSG is "aggressive" in its rendering in that there is a constant renderer loop running, making it easier to redraw upon all sorts of events such as resizing and user interaction. As part of the loop, the canvas element's properties are also set as needed. I am unsure if curve and especially rune have similar mechanisms that would enable similar behaviour, ie allow the canvas to both appear at a larger size *and* render at a larger resolution. Respective updates to the rendering mechanism as well as the WebGL portion that I think Ruo Mu wrote may be in order for this enhancement.

To summarise, here are some challenges that may need to be considered to make the curve canvas larger:
- Canvas element width/height properties are not updated dynamically by the renderer, causing the controls + flexbox to limit maximum canvas size to 512x512
- Renderer does not render at a dynamic/larger resolution depending on potentially available space for the canvas element
- Renderer is not triggered for a redraw when the tab (and thus potentially the canvas) resizes

Other possible considerations:
- Resolution/canvas behavioural consistency with the rune renderer

Apologies for the extremely long post, as I kept changing and adding things as my findings evolved. The post was originally meant to be much simpler. But in light of all this information, perhaps it will be fitting to "promote" this topic into its own standalone issue, as a more complex enhancement.

@martin-henz @sayomaki

**Comment by martin-henz** [comment]
Yes, the modules team is telling me the same. You may consider joining the discussion in telegram "SA 2425 Modules".

**Comment by sayomaki** [comment]
I think in my case I might have missed out on a styling property to set the width of the box containing the interface and model, and have tested modifications to my code to allow the box to expand fully to cover the space given (similar to runes/normal curve modules).

As what @Cloud7050 has mentioned, scaling up pixel-wise and ensuring consistency with other rendering modules that uses `WebGLCanvas` will definitely be more complicated, this seems more appropriate to be addressed in another separate issue instead.

For now, I have gone ahead to try to scale up both the slider and canvas output to the maximum dimensions available (slider will fill width while maintaining aesthetics, and canvas will scale up to the maximum viewport width/height as mentioned earlier), and I have a sample render of the new updated design.

<img width="1211" alt="New Updated Design" src="https://github.com/source-academy/modules/assets/26355099/2534adb5-d7e1-42f6-a53e-8a9bfe44f96f">
<br/><br/>
@martin-henz If this looks alright and well, I will proceed to draft a PR with the relevant changes and clean up the code that is going to be merged.

**Comment by martin-henz** [comment]
Yes, that's great! Please go ahead. Thanks! Looking forward to students winning the contest with this in place.
